sdk/src/main/scala/com/horizen/storage/SidechainStateStorage.scala

```diff
@@ -77,48 +81,129 @@ class SidechainStateStorage (storage : Storage, sidechainBoxesCompanion: Sidecha
     require(!boxIdsRemoveList.contains(null), "BoxId to remove must be NOT NULL.")
-    require(withdrawalRequestAppendList != null, "List of WithdrawalRequests to append must be NOT NULL. Use empty List instead.")
-    require(!withdrawalRequestAppendList.contains(null), "WithdrawalRequest to append must be NOT NULL.")
+    require(withdrawalRequestAppendSeq != null, "Seq of WithdrawalRequests to append must be NOT NULL. Use empty Seq instead.")
+    require(!withdrawalRequestAppendSeq.contains(null), "WithdrawalRequest to append must be NOT NULL.")
+    require(forgingStakesToAppendSeq != null, "Seq of ForgerStakes to append must be NOT NULL. Use empty Seq instead.")
+    require(!forgingStakesToAppendSeq.contains(null), "ForgerStake to append must be NOT NULL.")
 
     val removeList = new JArrayList[ByteArrayWrapper]()
     val updateList = new JArrayList[JPair[ByteArrayWrapper,ByteArrayWrapper]]()
 
+    // Update boxes data
     for (r <- boxIdsRemoveList)
       removeList.add(calculateKey(r))
 
     for (b <- boxUpdateList)
       updateList.add(new JPair[ByteArrayWrapper, ByteArrayWrapper](calculateKey(b.id()),
         new ByteArrayWrapper(sidechainBoxesCompanion.toBytes(b))))
 
-    updateList.add(new JPair(epochInformationKey,
+    // Update Withdrawal epoch related data
+    updateList.add(new JPair(withdrawalEpochInformationKey,
       new ByteArrayWrapper(WithdrawalEpochInfoSerializer.toBytes(withdrawalEpochInfo))))
 
-    if (withdrawalRequestAppendList.nonEmpty) {
+    if (withdrawalRequestAppendSeq.nonEmpty) {
       val withdrawalRequestList = getWithdrawalRequests(withdrawalEpochInfo.epoch)
 
-      withdrawalRequestList.addAll(withdrawalRequestAppendList.asJavaCollection)
+      withdrawalRequestList.addAll(withdrawalRequestAppendSeq.asJava)
 
       updateList.add(new JPair(getWithdrawalRequestsKey(withdrawalEpochInfo.epoch),
         new ByteArrayWrapper(withdrawalRequestSerializer.toBytes(withdrawalRequestList))))
-
     }
 
-    storage.update(version,
-      updateList,
-      removeList)
+    // Update Consensus related data
+    if(getConsensusEpoch.getOrElse(intToConsensusEpochNumber(0)) != consensusEpoch)
+      updateList.add(new JPair(consensusEpochKey, new ByteArrayWrapper(Ints.toByteArray(consensusEpoch))))
+
+    val (forgingStakesInfoSeq, forgingStakesAmount) = applyForgingStakesChanges(boxIdsRemoveList, forgingStakesToAppendSeq)
+    updateList.add(new JPair(
+      forgingStakesInfoKey,
+      new ByteArrayWrapper(forgingStakeInfoSerializer.toBytes(forgingStakesInfoSeq.asJava))
+    ))
+    updateList.add(new JPair(
+      forgingStakesAmountKey,
+      new ByteArrayWrapper(Longs.toByteArray(forgingStakesAmount))
+    ))
+
+    storage.update(version, updateList, removeList)
 
     this
   }
 
+  private def applyForgingStakesChanges(boxIdsToRemove: Set[Array[Byte]], forgingStakesToAppendSeq: Seq[ForgingStakeInfo]): (Seq[ForgingStakeInfo], Long) = {
```

Comment:
Do we need to restrict last ForgerBoxes spending? It will lead to possible chain stop in 2 epochs.
What is the impact to the snarks, if we will decide to add such a prevention rule?